Parse revert reason

karlb opened the pull request:
# What was wrong?

Solidity allows messages in `require` statements which can help a lot
when debugging. This commit parses the revert reasons from failing
`call`s and raises them as `SolidityError` exceptions.

Closes https://github.com/ethereum/web3.py/issues/941.

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)
- [x] Add unit test
- [x] Add integration test @kclowes 
- [x] Fix failures from integration test @marcgarreau 
- [x] Bump `eth-utils` (to account for the `replace_exceptions` change)

#### Cute Animal Picture

![IMG_20170706_123758](https://user-images.githubusercontent.com/144773/75170965-35708c00-572b-11ea-924e-47169a221284.jpg)

karlb commented on web3/manager.py:
```diff
@@ -150,6 +179,9 @@ def request_blocking(
 
         if "error" in response:
             apply_error_formatters(error_formatters, response)
```
Does this line have any effect? (This comment is unrelated to the changes in this PR)

karlb commented:
Is [this CI failure](https://app.circleci.com/jobs/github/ethereum/web3.py/98072) really caused by this PR? I don't see how the change could cause that.

kclowes commented:
@karlb no, that failure is flaky. Issue is [here](https://github.com/ethereum/web3.py/issues/1580). Feel free to ignore!

karlb commented:
> I think it would be good to add an integration test if you feel up to it, but I'm happy to add one if you don't.

@kclowes I'll gladly take you up on the offer of adding an integration test. I don't see how to fit it in nicely, right now.

I'll take care of the other change tomorrow.

kclowes commented:
Sounds good, thank you!

karlb commented:
I've moved the code to `method_formatters.py` and added a basic test for `get_error_formatters`. Did I get the change done the way you intended?

I'll probably have to fix something when there's a proper integration test, since I didn't test across different eth clients, yet.

kclowes commented:
@karlb I will keep digging into this tomorrow. I realized I read the original issue wrong and misdirected you towards putting the formatting in `method_formatters`. Based on the original issue, the formatting is correct when it goes through`eth_call`, but not when a function is called via `contract.functions.<function_name>.call()`. So I _think_ we'll need to do something around [here](https://github.com/ethereum/web3.py/blob/331f321835de5b842e0929dc6b416720634e46ee/web3/contract.py#L1433-L1437) to catch the `Revert` and re-raise a `SolidityError`, but I'm not 100% sure that is the right path. [This comment](https://github.com/ethereum/web3.py/issues/941#issuecomment-401422928) provides some direction as well. I will keep thinking about it and if you have any thoughts in the meantime, let me know.

I also don't think we need an integration test since `eth_call` is working as expected. I was able to write a quick contract with a revert function that you can use in the core tests to test the contract call interface. The commit is [here](https://github.com/ethereum/web3.py/commit/050bd27c36573b962d69d903a1723294034d265f), but eth-tester doesn't return the error message (it only returns the error type), so that will need to be addressed as well I think.

karlb commented:
> Based on the original issue, the formatting is correct when it goes through `eth_call`

It's not accidentally truncated. But I think web3py should detect reverts, parse the revert reason and raise that information as a proper Exception in that case, too. Getting
```python
SolidityError('not allowed to monitor'')
```
is much more useful than
```python
'Reverted 0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000166e6f7420616c6c6f77656420746f206d6f6e69746f7200000000000000000000'
```

> The commit is here, but eth-tester doesn't return the error message (it only returns the error type), so that will need to be addressed as well I think.

Yes, eth-tester is special. IIRC, it does not return errors as the other eth clients do, but directly raises python exceptions on transaction failures. This is usually very helpful, but for this specific PR it is causing additional work, since it is just an additional case to handle.

> I was able to write a quick contract with a revert function that you can use in the core tests to test the contract call interface.

Thanks! What is the easiest way (on a linux machine) to run that test trough geth and parity in addition to eth-tester (which seems to be what's used when I just use default pytest)?

kclowes commented:
> But I think web3py should detect reverts, parse the revert reason and raise that information as a proper Exception in that case, too.

Yes, I agree!

> it does not return errors as the other eth clients do, but directly raises python exceptions on transaction failures

In an effort to be backend-agnostic it changes errors that are specific to one EVM (like Revert) to a TransactionFailed error. 

> What is the easiest way (on a linux machine) to run that test trough geth and parity?

I think we'd have to do something like [this](https://github.com/ethereum/web3.py/blob/master/tests/generate_go_ethereum_fixture.py#L347) using the revert contract, then add it to the geth fixture at the end of the `setup_chain_state` method. After that, it could be added as a fixture in the respective conftest file (for example: https://github.com/ethereum/web3.py/blob/6ce485f554d13b93f43d9d7b415fbf158c65a77d/tests/integration/parity/conftest.py#L127), and then it could get used in the `module_testing` files. I will probably be able to get to that this week, or if you get to it first, that's great. Let me know if you need any more clarification!

karlb commented:
One sensible addition to the RevertContract could be a revert without message, so that we can test that case, too.

> In an effort to be backend-agnostic it changes errors that are specific to one EVM (like Revert) to a TransactionFailed error.

I'm now catching the eth-tester exception to create the result I intended. We'll see whether that lines up with the results from other backends.

I didn't find any time to look at setting up geth and parity. If you take care of that, I'll gladly have a look at any new failures.

kclowes commented:
> One sensible addition to the RevertContract could be a revert without message, so that we can test that case, too.

Good idea. I'll add that too. 

> I didn't find any time to look at setting up geth and parity. If you take care of that, I'll gladly have a look at any new failures.

Sounds good, thank you!

Exef commented:
Is it going to be merged soon?

If not @karlb, can I continue your work on this PR?

kclowes commented:
@Exef - @karlb was waiting on me to write some tests and then I lost track of this. Sorry about that! If you want to write some integration tests, please feel free!

karlb commented:
@Exef @kclowes Is one of you working on integration tests? I want to make sure that we don't accidentally have a deadlock where everyone waits for someone else to act next.

kclowes commented:
@karlb it's on my list, but I've been working on other things. I should be able to dig back in either later today or tomorrow but I'll let you know if that changes! And if @Exef gets to it first, that's great!

kclowes commented:
@karlb I just finished up the other PR that I was working on, so I should be able to get to this on Monday. Thanks for your patience!

kclowes commented:
@karlb I was able to make some progress on this today. I got the geth integration fixture up with the revert contract, and a simple test written, but I haven't had time to figure out why it is failing. I don't yet have the parity tests running either, but I should be able to keep working on it tomorrow. 

I also can't remember how I got code to you last time, so feel free to take or leave the commits I pushed to this branch!

karlb commented:
> @karlb I was able to make some progress on this today. I got the geth integration fixture up with the revert contract, and a simple test written, but I haven't had time to figure out why it is failing.

Unfortunately, I don't get the same error as in the CI locally:
```
pytest tests/integration/go_ethereum/test_goethereum_http.py::TestGoEthereumEthModuleTest::test_eth_call_revert_with_msg -x --tb=short
...
____________________________________________________________________________________________ TestGoEthereumEthModuleTest.test_eth_call_revert_with_msg[<lambda>] _____________________________________________________________________________________________
web3/_utils/module_testing/eth_module.py:664: in test_eth_call_revert_with_msg
    result = web3.codec.decode_single('bool', call_result)
venv/lib/python3.7/site-packages/eth_abi/codec.py:155: in decode_single
    return decoder(stream)
venv/lib/python3.7/site-packages/eth_abi/decoding.py:127: in __call__
    return self.decode(stream)
venv/lib/python3.7/site-packages/eth_abi/decoding.py:198: in decode
    raw_data = self.read_data_from_stream(stream)
venv/lib/python3.7/site-packages/eth_abi/decoding.py:308: in read_data_from_stream
    len(data),
E   eth_abi.exceptions.InsufficientDataBytes: Tried to read 32 bytes.  Only got 0 bytes
```
Full output: [error.txt](https://github.com/ethereum/web3.py/files/4735115/error.txt)

Apparently I get back `call_result = HexBytes('0x')`. Is my local setup broken, do I run the tests the wrong way, or is something else happening?

> I also can't remember how I got code to you last time, so feel free to take or leave the commits I pushed to this branch!

Please just push everything related into this branch!

pipermerriam commented on web3/manager.py:
```diff
@@ -146,10 +150,13 @@ def request_blocking(
-        response = self._make_request(method, params)
+        try:
+            response = self._make_request(method, params)
+        except eth_tester.exceptions.TransactionFailed as e:
```
I think in the best place to do this would be inside of the `EthereumTesterProvider` somewhere in here:

https://github.com/ethereum/web3.py/blob/master/web3/providers/eth_tester/main.py#L85-L103

That keeps us from having to have `eth-tester` specific code mixed into the main library code.

marcgarreau commented:
So I've been stepping through the eth_tester case first. Got stumped for a while before realizing the reason that the revert data never made it through to web3 was this `replace_exceptions` decorator; it swaps the revert out for an (empty) `TransactionFailed` exception:

https://github.com/ethereum/eth-tester/blob/51a208e02205639c37108873ccfddc4d845a0ddd/eth_tester/backends/pyevm/main.py#L488

Before I go yanking it out, is there anything I should know about the context around this exception handling? (It's not your code, but maybe you might know, @carver?)

carver commented:
> So I've been stepping through the eth_tester case first. Got stumped for a while before realizing the reason that the revert data never made it through to web3 was this `replace_exceptions` decorator; it swaps the revert out for an (empty) `TransactionFailed` exception:
> 
> https://github.com/ethereum/eth-tester/blob/51a208e02205639c37108873ccfddc4d845a0ddd/eth_tester/backends/pyevm/main.py#L488
> 
> Before I go yanking it out, is there anything I should know about the context around this exception handling? (It's not your code, but maybe you might know, @carver?)

Yeah, this all sounds right. The wrapping is because eth-tester wraps both pyethereum and py-evm, so we wanted to unify the exceptions in a way that was not tied to either back end.

I'm not totally sure we're ready to give that abstraction up. Maybe we can insert the wrapped exception's message in the abstract `TransactionFailed` exception?

marcgarreau commented:
Thanks for the tip, @carver. I think the right place to solve that is in eth-utils. Tiny PR up here: https://github.com/ethereum/eth-utils/pull/198.

marcgarreau commented:
Quick update: have still been digging around in the weeds here. Have learned that as of June, Geth returns revert messages in the JSON-RPC response ([v1.9.15](https://github.com/ethereum/go-ethereum/releases/tag/v1.9.15)). This has lead to some renewed conversation about which versions we should support in our test suite (https://github.com/ethereum/web3.py/issues/1703). If we only support the ~latest version(s), we may be able to forego some of the low-level decoding going on here. That decision/work may be considered a blocker to this PR.

karlb commented:
> Have learned that as of June, Geth returns revert messages in the JSON-RPC response ([v1.9.15](https://github.com/ethereum/go-ethereum/releases/tag/v1.9.15)). This has lead to some renewed conversation about which versions we should support in our test suite (#1703). If we only support the ~latest version(s), we may be able to forego some of the low-level decoding going on here.

Won't we need the decoding anyway to support revert messages for the OpenEthereum client?

marcgarreau commented:
> Won't we need the decoding anyway...?

Yeah, that stray idea was over-optimistic.

As for next steps, `call` (and the other `eth` namespaced methods) aren't getting the error formatters passed through, so the decoding logic isn't reached. Getting this PR over the line may be the next blocker: https://github.com/ethereum/web3.py/pull/1568/.

a0969818431 commented:
Close #1585

marcgarreau commented:
Got some cleanup to do, but at long last, I think the base functionality is there 🎉 

One stylistic choice to highlight: I applied Geth's convention to format all revert messages as `execution reverted: <reason>`.  I'm not attached to this idea; only priority is a standardizing across Geth, Parity/OE, and eth-tester. The alternatives would be to strip off the prefix from Geth or invent our own.

kclowes commented:
Awesome!!! I think Geth's convention is good, especially since we're raising the same `SolidityError` across the board. And I like that the revert word is in the message to give an extra clue.

marcgarreau commented:
Ready for review @kclowes (and anyone else interested). 

- More LOC than is ideal, but most are related to fixture generation with the revert contract and deduping two geth generation files. 
- Feedback especially welcome on the revert reason parsing code. I left most of Karl's original implementation to parse the raw data from Parity/OE. An alternative to that would be to use eth-abi's `decode_single` function, but IMO the readability is similar.

Thanks! 🙏

kclowes commented on tests/integration/test_ethereum_tester.py:
```diff
@@ -135,6 +135,29 @@ def txn_hash_with_log(block_with_txn_with_log):
+
+
+@pytest.fixture(scope="module")
+def revert_contract_address(revert_contract, address_conversion_func):
```
I don't see this getting used anywhere?

kclowes commented on web3/_utils/method_formatters.py:
```diff
@@ -486,6 +487,60 @@ def apply_list_to_array_formatter(formatter: Any) -> Callable[..., Any]:
 ABI_REQUEST_FORMATTERS = abi_request_formatters(STANDARD_NORMALIZERS, RPC_ABIS)
 
 
+def get_revert_reason(response: RPCResponse) -> str:
```
It feels like this `method_formatters` file is becoming a catch-all, so may be worth thinking about how we can split this out. But not for this PR :)

kclowes reviewed (approved):
Hooray! This will be so great to get in! I had mostly nitpicky comments and one about using `assert` in the code.

Also, because I can't comment on the file, I don't think we want to commit `tests/integration/tester.zip` :)

marcgarreau commented on tests/integration/generate_fixtures/common.py:
```diff
@@ -34,22 +34,28 @@
         "homesteadBlock": 0,
+        "byzantiumBlock": 0,
+        "constantinopleBlock": 0,
+        #  "petersburgBlock": 0,
```
I left these in because docs for syncing a parity + geth private net is woefully sparse. So, this is part documentation + part todo list. Does that context change anything or would you prefer that it live in an issue? No biggie either way.

kclowes replied:
Fair enough. I think it's fine here!

marcgarreau commented on tests/integration/generate_fixtures/common.py:
```diff
@@ -34,22 +34,28 @@
+        "0000000000000000000000000000000000000004": {"balance": "1"},
+        "0000000000000000000000000000000000000005": {"balance": "1"},
+        "0000000000000000000000000000000000000006": {"balance": "1"},
+        #  "0000000000000000000000000000000000000007": {"balance": "1"},
```
(will address in the same way as the note above)

marcgarreau commented on tests/integration/generate_fixtures/go_ethereum.py:
```diff
@@ -295,7 +176,7 @@ def verify_chain_state(web3, chain_data):
     web3.geth.miner.start(1)
     try:
-        return web3.eth.waitForTransactionReceipt(txn_hash, timeout=60)
+        return web3.eth.waitForTransactionReceipt(txn_hash, timeout=120)
```
Leaving it for good, since I regularly require it.

marcgarreau commented on tests/integration/generate_fixtures/common.py:
```diff
@@ -213,7 +219,7 @@ def mine_block(web3):
     start_time = time.time()
     web3.geth.miner.start(1)
-    while time.time() < start_time + 120:
+    while time.time() < start_time + 240:
```
this is prob overkill, will revert.

marcgarreau commented on web3/_utils/method_formatters.py:
```diff
@@ -486,6 +487,60 @@ def apply_list_to_array_formatter(formatter: Any) -> Callable[..., Any]:
+    See also https://solidity.readthedocs.io/en/v0.6.3/control-structures.html#revert
+    """
+    assert 'error' in response
+    if not isinstance(response['error'], dict):
```
Good question; not certain. Local linter is grumpy without it: `Item "str" of "Union[RPCError, str]" has no attribute "get"`. Will poke more.

marcgarreau replied:
So, I think web3 raises or prevent exceptions before reaching this point, but the RPCResponse type does include string values. I think leaving this guard clause won't do any harm, at least.

marcgarreau commented on tests/integration/test_ethereum_tester.py:
```diff
@@ -309,6 +332,12 @@ def test_eth_chainId(self, web3):
     def test_eth_getTransactionReceipt_mined(self, web3, block_with_txn, mined_txn_hash):
         super().test_eth_getTransactionReceipt_mined(web3, block_with_txn, mined_txn_hash)
 
+    def test_eth_call_revert_with_msg(self, web3, revert_contract, unlocked_account) -> None:
```
Unsure, don't think that was my doing. Tests pass without this code, so I'll remove.

sobolev-igor commented:
Hi everyone! 
Is there anything blocking this PR?

I'm having an issue connected to this PR (#1781) 
I really hope that this PR will fix the issue and I look forward to this PR being merged!

marcgarreau commented:
@sobolev-igor just leaving space for any final reviews. Expecting to merge later today or tomorrow, but next release is TBD.

njgheorghita commented on web3/_utils/method_formatters.py:
```diff
@@ -486,6 +487,58 @@ def apply_list_to_array_formatter(formatter: Any) -> Callable[..., Any]:
+
+def raise_solidity_error_on_revert(response: RPCResponse) -> RPCResponse:
+    revert_reason = get_revert_reason(response)
+    if revert_reason is None:
```
It seems to me that there is some unavoidable complexity here in dealing with the various clients, but in general return value checking is a bit of a code smell - one that I'm very familiar with cause I violate it constantly. But the [style guide](https://github.com/ethereum/snake-charmers-tactical-manual/blob/master/antipatterns.md#return-value-checking) goes into a bit more detail. One idea is to raise an exception in `get_revert_reason` if no reason is found and then handle the cases accordingly here in `raise_solidity_error_on_revert`.

njgheorghita commented on web3/_utils/module_testing/eth_module.py:
```diff
@@ -741,6 +742,38 @@ def test_eth_call_with_0_result(
+        revert_contract: "Contract",
+        unlocked_account: ChecksumAddress,
+    ) -> None:
+        with pytest.raises(SolidityError, match='execution reverted'):
```
Can we match a more specific part of the message here and the next test (like `execution reverted: not allowed to monitor` in `test_method_formatters.py`)?

njgheorghita commented on tests/core/utilities/test_method_formatters.py:
```diff
@@ -0,0 +1,65 @@
+})
+
+
+def test_get_revert_reason() -> None:
```
This test seems like it's worth [parameterizing](https://docs.pytest.org/en/stable/parametrize.html). Right now - for each successful test run we're testing 3 different case. Each test run should assert/disprove 1 case. 

Also the next test seems like it should be parametrized as well for the two cases that test if a `SolidityError` is raised, and then the `OTHER_ERROR` case split into its own test.

njgheorghita reviewed (approved):
Looks good to me! I just left a couple of thoughts on the style here & there, but nothing critical.

a0969818431 reviewed (commented):
`defi`

marcgarreau commented on tests/integration/generate_fixtures/go_ethereum.py:
```diff
@@ -363,6 +232,36 @@ def setup_chain_state(web3):
+    print('TXN_HASH_REVERT_WITH_NO_MSG:', txn_hash_revert_with_no_msg)
+    txn_receipt_revert_with_no_msg = common.mine_transaction_hash(web3, txn_hash_revert_with_no_msg)
+    block_hash_revert_no_msg = web3.eth.getBlock(txn_receipt_revert_with_no_msg['blockHash'])
+    print('BLOCK_HASH_REVERT_NO_MSG:', block_hash_revert_no_msg['hash'])
```
While generated fixtures, it has been nice to see these get printed in real time, because it fails/times out a nontrivial % of the time.

pipermerriam commented on web3/_utils/method_formatters.py:
```diff
@@ -486,6 +487,58 @@ def apply_list_to_array_formatter(formatter: Any) -> Callable[..., Any]:
+            return ''
+
+        reason_length = int(data[len(prefix):len(prefix) + 64], 16)
+        reason = data[len(prefix) + 64:len(prefix) + 64 + reason_length * 2]
```
This all feels fragile.  That might be "ok" but it still worries me.  Any thoughts on just doing the simple check that the data contains `"Reverted "` and then skipping this parsing logic and embedding the whole response message in the revert exception?

marcgarreau replied:
I might be misunderstanding, but I'd definitely like to include the parsed revert message and not put that on the end user's shoulders. Are you suggesting returning both the reason and the full error dict, just the error dict, or something else?

marcgarreau replied:
Current thinking in support of :shipit: : 
- in related [issues filed](https://github.com/ethereum/web3.py/search?q=revert&type=issues), users are generally looking for the plaintext reason.
- handling reverts is broken today, so even brittle parsing logic is an improvement. 
- I don't expect clients to change this formatting on a whim. 
- we can iterate on it.

pipermerriam replied:
I'm proposing that we return the string which we are trying to parse in full.  basically this whole context block would become:

```python
if data.startswith("Reverted "):
    reason = data
```

I don't feel strongly about this.  It just seems like it would be roughly the same usefulness to the user and would avoid the part that looks brittle.

marcgarreau replied:
Makes good sense, but it feels like signing up for an endless trickle of "what is this? how do I parse this?" questions. I'd prefer to sign up for patching the logic on the hopefully very rare occasions required and deliver the superior UX. Will flag as this something "to harden" though.

pipermerriam commented on web3/_utils/method_formatters.py:
```diff
@@ -486,6 +487,58 @@ def apply_list_to_array_formatter(formatter: Any) -> Callable[..., Any]:
+    if 'message' in response['error'] and response['error'].get('code', '') == 3:
+        return response['error']['message']
+
+    return None
```
gentle nudge to [avoid patterns that require return value checking](https://github.com/ethereum/snake-charmers-tactical-manual/blob/master/antipatterns.md#return-value-checking)

pipermerriam replied:
Also, `mypy` should be yelling at you about this since this function is designated to return a `str` so it is probably worth looking into why those checks aren't being enforced.

kclowes commented on tests/core/utilities/test_method_formatters.py:
```diff
@@ -0,0 +1,78 @@
+        'test-get-revert-reason-without-msg',
+    ])
+def test_get_revert_reason(response, expected) -> None:
+    with pytest.raises(SolidityError):
```
This is pretty nitpicky, but that assert will never get hit. I think this would be better like:
```
    with pytest.raises(SolidityError, match=expected):
        raise_if_revert_reason(response)
```

fubuloubu commented on web3/exceptions.py:
```diff
@@ -192,3 +192,11 @@ class InvalidEventABI(ValueError):
     pass
+
+
+class SolidityError(ValueError):
+    # Inherits from ValueError for backwards compatibility
+    """
+    Raised on a solidity require/revert
+    """
+    pass
```
This seems like a breaking change to the API of expected failures for web3. What was the reasoning for intercepting `TransactionFailed` failures and converting to this error (which also seems poorly named, perhaps `AssertionError(TransactionFailed)` would be better considering Solidity isn't the only smart contract language out there that has a assertion feature)
